### PR TITLE
Add support for memory mapping of NdArray chunks

### PR DESCRIPTION
**Pull request by danieldk**

### CODE REVIEW

File: rust2vec/src/storage.rs [hunk 1]
```diff
@@ -47,39 +49,127 @@ pub type CowArray1<'a, A> = CowArray<'a, A, Ix1>;
...
+        write.write_u32::<LittleEndian>(f32::type_id())?;
+
+        // Write padding, such that the embedding matrix starts on at
+        // a multiple of the size of f32 (4 bytes). This is necessary
+        // for memory mapping a matrix. Interpreting the raw u8 data
+        // as a proper f32 array requires that the data is aligned in
+        // memory. However, we cannot always memory map the starting
+        // offset of the matrix directly, since mmap(2) requires a
+        // file offset that is page-aligned. Since the page size is
+        // always a larger power of 2 (e.g. 2^12), which is divisable
```
Comment (sebpuetz) [inline rust2vec/src/storage.rs]: `divisible` ;)

File: rust2vec/src/storage.rs [hunk 2]
```diff
@@ -47,39 +49,127 @@ pub type CowArray1<'a, A> = CowArray<'a, A, Ix1>;
...
+        // by 4, the offset of the matrix with regards to the page
+        // boundary is also a multiple of 4.
+
+        let n_padding = padding::<f32>(write.seek(SeekFrom::Current(0))?);
+        let padding = vec![0; n_padding as usize];
+        write.write(&padding)?;
+
```
Comment (sebpuetz) [inline rust2vec/src/storage.rs:117]: This comment is to verify that I'm correctly reading this method:
`write.seek(SeekFrom::Current(0))?);` returns the current position of the writer and `padding::<f32>(position)` then returns the bytes missing to hit the alignment to multiples of `f32`.

File: rust2vec/src/storage.rs [hunk 3]
```diff
@@ -47,39 +49,127 @@ pub type CowArray1<'a, A> = CowArray<'a, A, Ix1>;
...
+        let shape = Ix2(rows, cols);
+
+        ensure!(
+            read.read_u32::<LittleEndian>()? == f32::type_id(),
+            "Expected single precision floating point matrix for NdArray."
+        );
+
+        let n_padding = padding::<f32>(read.seek(SeekFrom::Current(0))?);
+        read.seek(SeekFrom::Current(n_padding as i64))?;
+
```
Comment (sebpuetz) [inline rust2vec/src/storage.rs:149]: Same idea as above, just verifying I'm understanding the logic here:
`read.seek(SeekFrom::Current(0))?)` returns the position of the reader after reading the known parts. In theory the `seek` call could be replaced by adding up the width of the previously read items, although that would make maintenance more bug-prone, if the header were ever subject to change.

**Review by sebpuetz — APPROVED** [review]
Only found an orthographic error. Rest of the comments are just there to verify that I didn't misunderstand what's going on.

File: rust2vec/src/storage.rs (hunk 2 above)
Comment (danieldk) [reply rust2vec/src/storage.rs:117]: Correct on both accounts.

File: rust2vec/src/storage.rs (hunk 3 above)
Comment (danieldk) [reply rust2vec/src/storage.rs:149]: In theory yes, and not only the header but also all preceding chunks. (But I think you meant that.) Lots of bookkeeping :).

**Pull request merged**